sw/device/lib/sdk/dma/dma_sdk.c

```diff
@@ -16,835 +16,63 @@
-                   DMA_SRC_DATA_TYPE_DATA_TYPE_MASK,
-                   DMA_SELECTION_OFFSET_START,
-                   peri);
+        /* Enable fast interrupts */
+        CSR_SET_BITS(CSR_REG_MIE, DMA_CSR_REG_MIE_MASK);
 
-    write_register(trans.dst_type,
-                   DMA_DST_DATA_TYPE_REG_OFFSET,
-                   DMA_DST_DATA_TYPE_DATA_TYPE_MASK,
-                   DMA_SELECTION_OFFSET_START,
-                   peri);
-
-    peri->INTERRUPT_EN = 0x1;
-
-    /* Load the size and start the transaction. */
-    peri->SIZE_D1 = size * sizeof(uint8_t);
-
-    #endif
-
-    while (!dma_is_ready(channel))
-    {
-        // disable_interrupts
-        // this does not prevent waking up the core as this is controlled by the MIP register
-        CSR_CLEAR_BITS(CSR_REG_MSTATUS, 0x8);
-        if (dma_is_ready(channel) == 0)
-        {
-            wait_for_interrupt();
-            // from here we wake up even if we did not jump to the ISR
-        }
-        CSR_SET_BITS(CSR_REG_MSTATUS, 0x8);
+        return;
     }
 
-    return;
-}
-
-
-
-void dma_copy_16_32(uint32_t *dst, uint16_t *src, uint32_t size, uint8_t channel)
-{
-
-    dma *peri = dma_peri(channel);
-
-    uint8_t dataSize_b = DMA_DATA_TYPE_2_SIZE(DMA_DATA_TYPE_WORD);
-
-    /*
-     * SET THE POINTERS
-     */
-    peri->SRC_PTR = (uint32_t) src;
-    peri->DST_PTR = (uint32_t) dst;
-
-    /*
-     * SET THE INCREMENTS
-     */
-
-    write_register(2,
-                   DMA_SRC_PTR_INC_D1_REG_OFFSET,
-                   DMA_SRC_PTR_INC_D1_INC_MASK,
-                   DMA_SRC_PTR_INC_D1_INC_OFFSET,
-                   peri);
-
-    write_register(4,
-                   DMA_DST_PTR_INC_D1_REG_OFFSET,
-                   DMA_DST_PTR_INC_D1_INC_MASK,
-                   DMA_DST_PTR_INC_D1_INC_OFFSET,
-                   peri);
-    /*
-     * SET THE OPERATION MODE AND WINDOW SIZE
-     */
-
-    peri->MODE = DMA_TRANS_MODE_SINGLE;
-
-    write_register(DMA_DATA_TYPE_HALF_WORD,
-                   DMA_SRC_DATA_TYPE_REG_OFFSET,
-                   DMA_SRC_DATA_TYPE_DATA_TYPE_MASK,
-                   DMA_SELECTION_OFFSET_START,
-                   peri);
-
-    peri->INTERRUPT_EN = 0x1;
-
-    /* Load the size and start the transaction. */
-    peri->SIZE_D1 = dataSize_b * size;
-
-    // #endif
-
-    while (!dma_is_ready(channel))
+    void dma_copy(uint32_t dst_ptr, uint32_t src_ptr, uint32_t size, uint8_t channel, dma_data_type_t src_type, dma_data_type_t dst_type, uint8_t signed_data)
     {
-        // disable_interrupts
-        // this does not prevent waking up the core as this is controlled by the MIP register
-        CSR_CLEAR_BITS(CSR_REG_MSTATUS, 0x8);
-        if (dma_is_ready(channel) == 0)
-        {
-            wait_for_interrupt();
-            // from here we wake up even if we did not jump to the ISR
-        }
-        CSR_SET_BITS(CSR_REG_MSTATUS, 0x8);
+        volatile dma *the_dma = dma_peri(channel);
```

Comment:
1) `the_dma` ? hahaha 
2) If `the_dma` is being enforced to be volatile, having a separate function to start it does not make much sense. If it's only for clarity, maybe consider using `__attribute__((always_inline))` ?

Reply:
if we are having a separate function, maybe there is no need to have the volatile, allowing the compiler to accelerate a bit the copying of the data. 

I think its a small detail. For what it concerns me, it can be left like this (maybe not with `the_dma` tho hahaha)